### PR TITLE
Windows: Rename layers before deletion

### LINKED ISSUE #27570: Windows: layers can be partially removed, causing unpredictable failures
The Windows graph driver tries to delete a layer's state by recursively removing files from within the layer's directory. If this fails partway through, the layer is not removed from the layer DB, but the layer contents are also not restored. So if the layer is referenced again by another image, starting a container based on that image will fail in unpredictable ways.

The aufs graph driver avoids this problem by, as a first step, renaming the layer directory to append "-removing", and does not report failure if there are any unlink failures while removing the layer contents. The Windows graph driver should probably do the same.

It looks to me like Linux might have this problem too, for the overlay and overlay2 graph drivers.

@jhowardmsft, @darrenstahlmsft, @swernli

### PR DESCRIPTION
**Pull request by darstahl**
fixes https://github.com/docker/docker/issues/27570

**- What I did**

Layers are less likely to enter a partially removed state.

**- How I did it**

By renaming the layers to *-removing before removal starts, the rename will fail if the layer is in use, preventing the partial deletion by DestroyLayer.

**- How to verify it**

Manual testing that when a container is leaked, the layer is not partially removed, but instead the rename fails.

Signed-off-by: Darren Stahl <darst@microsoft.com>

### CODE REVIEW

**Comment by MHBauer** [comment]
That seems sane assuming you can't rename an in-use file. Is there any documentation that says so? I am curious more than anything.

Code looks good.

**Comment by darrenstahlmsft** [comment]
@MHBauer It depends on the flags passed to OpenFile/CreateFile. On Windows, you must request the ability for shared deletion with the FILE_SHARED_DELETE flag at creation, otherwise it cannot be renamed or deleted by other processes. Containers intentionally do not use the FILE_SHARED_DELETE flag, as a result, a rename will fail if the container is in use. There's still the chance for something to grab a file between the rename and the delete, but the layerstore should no longer see the leaked files since they have the -removing in the layer name.

See https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx

Thinking about it more, the DestroyLayer errors should be ignored since the folder is already renamed, and we can't do anything about it at that point. It should also be removed from the layer store if the DestroyLayer fails. I've updated the PR slightly.

File: daemon/graphdriver/windows/windows.go [hunk 1]
```diff
@@ -244,7 +244,18 @@ func (d *Driver) Remove(id string) error {
 		return err
 	}
 	os.RemoveAll(filepath.Join(d.info.HomeDir, "sysfile-backups", rID)) // ok to fail
-	return hcsshim.DestroyLayer(d.info, rID)
+
+	layerPath := path.Join(d.info.HomeDir, rID)
+	tmpID := fmt.Sprintf("%s-removing", rID)
+	tmpLayerPath := path.Join(d.info.HomeDir, tmpID)
```
Comment (cpuguy83) [inline daemon/graphdriver/windows/windows.go]: How come you are using `path` instead of `filepath`. Won't this create a bunch of `/`?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: daemon/graphdriver/windows/windows.go (hunk 1 above)
Comment (darrenstahlmsft) [inline daemon/graphdriver/windows/windows.go]: Ah! That's where the slashes were coming from. Fixed.

**Review by cpuguy83 — APPROVED** [review]
LGTM

File: daemon/graphdriver/windows/windows.go [hunk 2]
```diff
@@ -244,7 +244,18 @@ func (d *Driver) Remove(id string) error {
 		return err
 	}
 	os.RemoveAll(filepath.Join(d.info.HomeDir, "sysfile-backups", rID)) // ok to fail
```
Comment (jstarks) [inline daemon/graphdriver/windows/windows.go]: I believe you can remove this line now. I don't see where we ever populate this directory.

Comment (darrenstahlmsft) [inline daemon/graphdriver/windows/windows.go]: Yep, removed it.

**Comment by jstarks** [comment]
LGTM

**Comment by jhowardmsft** [comment]
LGTM. Merging.

**Pull request merged**